### ISSUE TITLE
"dataurl" transformer outputs invalid image

### ISSUE THREAD

**Issue by mnghsn** [eight04/inline-js#4]
The `dataurl` transformer can't output correct base64 string if input file is binary.

```js
$inline('whatever.png|dataurl:image/png')
// -> data:image/png;charset=utf8;base64,77+9UE5HDQoaCgAAAA1JSERSAAAAEAAAABAIBgAAAB/vv...
```

Browsers can't load this data URI as a PNG image. A correct one should be:

```
data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAA8ElEQVR42o...
```

(node@8.5.0, npm@5.4.2, Ubuntu 14.04)

**Comment by eight04** [comment]
https://github.com/eight04/inline-js/blob/13de91013a7d0103390b5ed67d85b09ee0d7ef7d/.inline.js#L40-L50

I guess we need to make charset default to binary.

**Comment by eight04** [comment]
Fixed in 0.4.0.

**Comment by mnghsn** [comment]
It seems not working properly because `fs.readFileSync` uses `utf8` encoding.

https://github.com/eight04/inline-js/blob/283078957634c6c545ed21f327feca9a97252b56/.inline.js#L2-L15

**Comment by eight04** [comment]
Yeah the fix doesn't work since the content is already broken when converting the buffer into utf8 string.

We have to change how we handle content entirely and provide a way to read raw file. Some thoughts:

1. Try to detect if the file is binary (https://www.npmjs.com/package/is-binary-path), and decide if it should be convert into utf8 string. If failed to detect, read as utf8 string by default.
2. Provide a way to explicitly read raw/text file. Maybe `$inline("raw:myfile")` and `$inline("text:myfile")`.
3. Don't search `$inline` tag for a binary file, only apply transformers. This change should be safe since no one puts `$inline` tag in a binary file.
4. If the result of the transforms is a `Buffer`, convert it into binary string.

**Comment by eight04** [comment]
Fixed in 0.5.0. Now inline-js can correctly read/write binary file.
